dkg/dkg.go

```diff
@@ -77,16 +77,30 @@ func (c Config) HasTestConfig() bool {
 //nolint:maintidx // Refactor into smaller steps.
 func Run(ctx context.Context, conf Config) (err error) {
 	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
 
 	ctx = log.WithTopic(ctx, "dkg")
 
 	lockSvc, err := privkeylock.New(p2p.KeyPath(conf.DataDir)+".lock", "charon dkg")
 	if err != nil {
+		// cancel manually here because we'll defer cancel() and lockSvc.Done() later
+		cancel()
 		return err
 	}
 
+	// Make sure to always wait for lockSvc to be done.
+	lockSvcDone := make(chan struct{})
+
+	defer func() {
+		// Explicitly cancel the context and wait until the privkey lock is deleted.
+		cancel()
+		<-lockSvcDone
+	}()
+
 	go func(ctx context.Context) {
+		defer func() {
+			lockSvcDone <- struct{}{}
+		}()
```

Comment:
can just do:
```suggestion
		defer close(lockSvcDone)
```

Reply:
It's pretty much the same thing, `struct{}` has zero size in memory.